master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+		delayBytes: delayBytes,
+		buf:        make([]byte, 0, delayBytes),
+		next:       w,
+	}
+}
+
+// S3SeqWriterAt satisfies S3 APIs' io.WriterAt interface while staying sequential.
+// Ref: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#Downloader
+type s3SeqWriterAt struct {
+	next    io.Writer
+	written int64
+}
+
+func newS3SeqWriterAt(w io.Writer) *s3SeqWriterAt {
+	return &s3SeqWriterAt{next: w}
+}
+
+// WriteAt writes the content in buffer p.
+func (w *s3SeqWriterAt) WriteAt(p []byte, off int64) (int, error) {
+	if off != w.written {
+		return 0, fmt.Errorf(
+			"only supporting sequential writes,"+
+				" writing at offset %d while %d bytes have been written",
+			off, w.written)
+	}
+	n, err := w.next.Write(p)
+	if err != nil {
+		return 0, err
+	}
+	w.written += int64(n)
+
+	return n, err
+}
+
+// BatchDownloadIterator implements s3's BatchDownloadIterator API.
+type batchDownloadIterator struct {
+	// The objects we are writing
+	objects []*s3.Object
+	// The output we are writing to
+	aw archiveWriter
+	// Internal states
+	err    error
+	pos    int
+	bucket string
+}
+
+// Next() returns true if the next item is available.
+func (i *batchDownloadIterator) Next() bool {
+	i.pos++
+	if i.pos == len(i.objects) {
+		return false
+	}
+	err := i.aw.WriteFileHeader(*i.objects[i.pos].Key, *i.objects[i.pos].Size)
+	if err != nil {
+		i.err = err
+		return false
+	}
+	return true
+}
+
+// Err() eturns the error if any.
+func (i *batchDownloadIterator) Err() error {
+	return i.err
+}
+
+// DownloadObject() eturns a DownloadObject.
+func (i *batchDownloadIterator) DownloadObject() s3manager.BatchDownloadObject {
+	return s3manager.BatchDownloadObject{
+		Object: ptrs.Ptr(s3.GetObjectInput{
+			Bucket: ptrs.Ptr(i.bucket),
+			Key:    i.objects[i.pos].Key,
+		}),
+		Writer: newS3SeqWriterAt(i.aw),
+	}
+}
+
+func newBatchDownloadIterator(aw archiveWriter,
+	bucket string, objs []*s3.Object) *batchDownloadIterator {
+	return &batchDownloadIterator{
+		aw:      aw,
+		bucket:  bucket,
+		objects: objs,
+		pos:     -1,
+	}
+}
+
+func getAwsRegion() string {
+	defaultRegion := "us-west-2"
+	client := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+
+	// Per AWS: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	resp, err := client.Get("http://169.254.169.254/latest/dynamic/instance-identity/document")
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return defaultRegion
+	}
+
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return defaultRegion
+	}
+	_ = resp.Body.Close()
+
+	var jsonObj map[string]interface{}
+	err = json.Unmarshal(bytes, &jsonObj)
+	if err != nil {
+		return defaultRegion
+	}
+
+	if _, ok := jsonObj["region"]; !ok {
+		return defaultRegion
+	}
+
+	region, ok := jsonObj["region"].(string)
+	if !ok {
+		return defaultRegion
+	}
+
+	return region
+}
+
+func s3DownloadCheckpoint(
+	c echo.Context, aw archiveWriter, bucket string, prefix string,
```

Comment:
is it possible to accept a generic `context.Context` here instead of the `echo.Context`?

I'm thinking accepting an echo.Context makes this way harder to unit test.

Reply:
Seems doable. In a related question,  for unit tests, how much testing is preferred -- I will also add integration tests.?

Reply:
if it were me, I would only worry about testing against the real s3 api; I wouldn't bother with any fully self-contained unit tests.